lib/src/facts/linux/operating_system_resolver.cc

```diff
@@ -21,82 +22,186 @@ namespace bs = boost::system;
+                if (re_search(line, boost::regex("(?m)^(\\w+)=[\"']?(.+?)[\"']?$"), &key, &value)) {
+                    values.insert(make_pair(key, value));
+                }
                 return true;
-            }
-            *variable = line.substr(offset);
-            boost::trim(*variable);
-            return true;
-        });
+            });
+        }
+        return values;
+    }
 
-        auto name = get_name(result.distro.id);
-        if (!name.empty()) {
-            result.name = move(name);
+    struct os_linux
+    {
+        os_linux(map<string, string> release_info) : _release_info(move(release_info)) {}
+        virtual string get_name(string const& distro_id);
+        virtual string get_family(string const& name);
+        virtual string get_release(string const& name, string const& distro_release);
+     protected:
+        map<string, string> _release_info;
+    };
+
+    struct os_cumulus : os_linux
+    {
+        os_cumulus(map<string, string> release_info) : os_linux(move(release_info)) {}
+        virtual string get_name(string const& distro_id) override
+        {
+            return os::cumulus;
+        }
+        virtual string get_family(string const& name) override
+        {
+            return os_family::debian;
+        }
+        virtual string get_release(string const& name, string const& distro_release) override
+        {
+            return _release_info["VERSION_ID"];
         }
+    };
 
-        auto release = get_release(result.name, result.distro.release);
-        if (!release.empty()) {
-            result.release = move(release);
+    struct os_coreos : os_linux
+    {
+        os_coreos(map<string, string> release_info) : os_linux(move(release_info)) {}
+        virtual string get_name(string const& distro_id) override
+        {
+            return os::coreos;
+        }
+        virtual string get_family(string const& name) override
+        {
+            return os_family::coreos;
+        }
+        virtual string get_release(string const& name, string const& distro_release) override
+        {
+            return _release_info["VERSION_ID"];
         }
+    };
 
-        // Convert the architecture value depending on distro
-        // For certain distros, use "amd64" for "x86_64"
-        // For certain distros, use "x86" for "i386"
-        if (result.architecture == "x86_64" && (
-             result.name == os::debian ||
-             result.name == os::gentoo ||
-             result.name == os::kfreebsd ||
-             result.name == os::ubuntu)) {
-            result.architecture = "amd64";
-        } else if (re_search(result.architecture, boost::regex("i[3456]86|pentium"))) {
-            // For 32-bit, use "x86" for Gentoo and "i386" for everyone else
-            if (result.name == os::gentoo) {
-                result.architecture = "x86";
-            } else {
-                result.architecture = "i386";
+    static os_linux get_os()
```

Comment:
Does this work?  We're returning an `os_linux` and not a pointer/reference type.